NEO-2115 // update Table row selection

joe-s-avaya opened the pull request:
[Link to updated Table story](https://deploy-preview-444--neo-react-library-storybook.netlify.app/?path=/story/components-table--editable-data)

**Before tagging the team for a review, I have done the following:**

- [x] run `yarn all` locally: ensures that all tests pass, formatting is done, types pass, and builds pass
- [x] reviewed my code changes
- [x] updated the link at the top of this PR (or remove it if not applicable)
- [x] added any important information to this PR (description, images, GIFs, ect.)
- [x] done an accessibility check on my work (tested with Chrome's `axe Dev Tools`, Mac's VoiceOver, etc.)
- [x] tagged `@avaya-dux/dux-design` if any visual changes have occurred
- [x] tagged `@avaya-dux/dux-devs`

This makes several updates to row selection in the Table component. Most notably that the header checkbox now selects/deselects the _page's_ rows, _not_ the entire tables.

[Figma link](https://www.figma.com/file/o62kJbRQpFq9T0fGwKvCMA/Component-Exploration-FY22---v3?type=design&node-id=3313-531108&mode=design&t=1IRQvmML9xvf4Oo0-0)

`@avaya-dux/dux-design`: Review table selection visuals.

- "Select all" checkbox appropriately appears and updates.
- Chevron next to the checkbox is visually appropriate (sizing, spacing, etc.).

`@avaya-dux/dux-devs`: Review table selection logic.

- review code and test logic
- review [use-cases spreadsheet](https://avaya365-my.sharepoint.com/:x:/g/personal/jsebast_avaya_com/EaYuTO3lFF9DiKXE1jcrCPcBR7VnZojp50o4zXPv_JNy1Q?e=Q8jXVg) and ensure all use-cases are handled
- review updates to Menu components and stories to ensure nothing has inappropriately changed. 
  - IMPORTANT: all current "Menu Item" components will not be fully a11y due to them not being `<button>`s. We'll need to figure out how to fix that soon.




<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced row selection functionality in tables, including selecting visible rows, clearing selections, and toggling row states.
- **Improvements**
  - Updated table header translations for better clarity on selection actions.
- **Bug Fixes**
  - Simplified row selection logic in test cases to improve reliability.
  - Improved checkbox label references for better test accuracy.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yangAtSpoken commented:
<img width="785" alt="Screenshot 2024-07-09 at 1 14 10 PM" src="https://github.com/avaya-dux/neo-react-library/assets/31546063/66c20c08-065d-4e55-a05a-a6b8c2b77292">

Use Case review:  the last use case does not seem working. My understand of this use case is: when all rows cross all pages are selected, clicking a header checkbox of any page, will de-select all items cross all pages.  What actually happens is  only rows of the clicked page are deselected.

yangAtSpoken commented:
<img width="271" alt="functional" src="https://github.com/avaya-dux/neo-react-library/assets/31546063/a3e187de-8c42-447e-a50a-1787cba49b5a">

Can not move focus to the last menu item with keyboard.

yangAtSpoken commented on src/components/Table/Table.test.jsx:
```diff
@@ -315,16 +312,21 @@ describe("Table", () => {
+
+			// delete second (final) page
+			await user.click(screen.getByLabelText(selectAllCheckboxLabel));
+			await user.click(screen.getByText("Delete"));
```
nice

yangAtSpoken commented on src/components/Table/TableComponents/TableHeader.tsx:
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
 		toggleSortBy,
 	} = instance;
 
+	const [visibleRows, allVisibleRowsAreSelected, allVisibleRowsDeselected] =
```
"visible" and "shown" are not accurate to me. If the page is not selected, all rows in that page are not visible/shown to the user.

joe-s-avaya replied:
Excellent call! As they say, naming is the hardest problem in software, and your recommendations are a significant improvement; thanks for being thorough! I have made the updates that you recommended.

yangAtSpoken commented on src/components/Table/TableComponents/TableHeader.tsx:
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
 
+	const [visibleRows, allVisibleRowsAreSelected, allVisibleRowsDeselected] =
+		useMemo(() => {
+			const shownRows = page.map((row) => row.original);
```
shownRows => pageRows

joe-s-avaya replied:
updated

yangAtSpoken commented on src/components/Table/TableComponents/TableHeader.tsx:
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
+	const [visibleRows, allVisibleRowsAreSelected, allVisibleRowsDeselected] =
+		useMemo(() => {
+			const shownRows = page.map((row) => row.original);
+			const visibleRowsSelected =
```
allRowsSelected or allPageRowsSelected

joe-s-avaya replied:
updated

yangAtSpoken commented on src/components/Table/TableComponents/TableHeader.tsx:
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
+			const shownRows = page.map((row) => row.original);
+			const visibleRowsSelected =
+				shownRows.length && shownRows.every((row) => selectedRowIds[row.id]);
+			const visibleRowsNotSelected = shownRows.every(
```
allRowsNoteSlected or noRowsSelected

joe-s-avaya replied:
updated

yangAtSpoken commented on src/components/Table/TableComponents/TableHeader.tsx:
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
+			return [shownRows, visibleRowsSelected, visibleRowsNotSelected];
+		}, [page, selectedRowIds]);
+
+	const selectVisibleRows = useCallback(
```
selectPageRows

joe-s-avaya replied:
updated

yangAtSpoken commented on src/components/Table/TableComponents/TableHeader.tsx:
```diff
@@ -32,13 +32,35 @@ export const TableHeader = <T extends Record<string, any>>({
+		}, [page, selectedRowIds]);
+
+	const selectVisibleRows = useCallback(
+		(set: boolean) => {
```
selected: bolean

joe-s-avaya replied:
updated

yangAtSpoken commented on src/components/Table/TableComponents/TableHeader.tsx:
```diff
@@ -69,19 +91,73 @@ export const TableHeader = <T extends Record<string, any>>({
 		checkboxCheckedValue,
 		rowsById,
 	]);
+	const toggleAllVisibleRows = useCallback(() => {
```
togglePageRows

joe-s-avaya replied:
updated

yangAtSpoken reviewed (approved):
The Table update looks great to me despite of all the naming suggestions.  However, I do not see that we need a MenuItemButton component as MenuItem has onClick callback already.

joe-s-avaya commented on src/components/Menu/MenuItemButton/MenuItemButton.tsx:
```diff
@@ -0,0 +1,19 @@
+	children?: React.ReactNode;
+}
+
+export const MenuItemButton = ({
```
The original `MenuItem` is a `div`, which does not have the built-in a11y that the `button` has. However, due to our `Menu` assuming that we're handling all the things, I'm running into quite a few issues with my implementation and may need to throw all of this away 😭 

An example of the built-in a11y is that a `button`'s `onClick` is automatically triggered by the user hitting "enter" or "space" (and other inputs for screen readers). However, the `Menu` is forcibly stopping propagation, disrupting how the base `button` works. So, I'm going to play with it a bit more, but I think I should throw this new component.

Justin-Avaya commented:
** Possibly out of scope
Any reason for three different hover treatments?
- Hover over Select all section is a slow fade in (too slow in my opinion)
- Hover over rows seems instant
- Hover over column header names is a sweeping animation

Example:
https://github.com/avaya-dux/neo-react-library/assets/161497909/45ce25a9-5010-42d1-927e-3910e9a60bd3

<hr>

** Possibly out of scope
As the rows are selected with the Select All checkbox, the columns shift/jump around. Selecting individual rows does not seem to have this effect with the exception of selecting the 4th row (Madam Hailey) which also creates the same shifting/jumping. This could be a row issue and not the Select All feature

Example:
https://github.com/avaya-dux/neo-react-library/assets/161497909/8540692f-ec2e-44eb-8f65-e725ccfc02ba

<hr>

Vertical sizing seems off from the source. The Figma document indicates the vertical height should be 44px whereas the story shows a vertical height of 61px

Example:
<img width="314" alt="Screenshot 2024-07-09 at 12 41 05 PM" src="https://github.com/avaya-dux/neo-react-library/assets/161497909/71bbb8f6-8c8c-409e-a9c3-8d0e3a0fc2c2">

<img width="625" alt="Screenshot 2024-07-09 at 12 43 10 PM" src="https://github.com/avaya-dux/neo-react-library/assets/161497909/f2280d3b-ceca-4af9-ac13-b25f26c336a9">

<hr>

If all of the above are out of scope and need to be addressed elsewhere, just let me know. The rest of it looked fine.

joe-s-avaya commented:
Need to do some refactoring, closing for now and will re-open.

joe-s-avaya commented:
> **Possibly out of scope**

@Justin-Avaya, this is all out of scope for this PR as my changes did not cause this. However, this many inconsistencies is particularly worrying and I've thus made a [jira issue](https://jira.forge.avaya.com/browse/NEO-2316) to resolve these. I'll bring this up with the team in standup tomorrow so we can schedule it.

joe-s-avaya commented:
> <img alt="functional" width="271" src="https://private-user-images.githubusercontent.com/31546063/347056412-a3e187de-8c42-447e-a50a-1787cba49b5a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1NDkxMjQsIm5iZiI6MTcyMDU0ODgyNCwicGF0aCI6Ii8zMTU0NjA2My8zNDcwNTY0MTItYTNlMTg3ZGUtOGM0Mi00NDdlLWE1MGEtMTc4N2NiYTQ5YjVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA5VDE4MTM0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlZTU5OTA5M2I1OTJjMWYzNjZmZjllNjFjODg2NmQ4MDI4ZmMzMTlhNTNmMzQ3MWViMjUxNGUxNzEzM2I2OTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yAXQKxyMLNT7x0bMkJVHQ3Wr8iObTlWWaNeLyUpNJuI">
> Can not move focus to the last menu item with keyboard.

@yangAtSpoken can't reproduce. I must have broken it and then fixed it all without realizing it 😅

Justin-Avaya reviewed (approved):
- [x] "Select all" checkbox appropriately appears and updates.
- [x] Chevron next to the checkbox is visually appropriate (sizing, spacing, etc.).

✅ Approved

enrique-prado commented:
From a UX perspective I think we are missing a "deselect all rows" option. If the user selects all rows and then decides to unselect them then he/she is forced to go to every page and deselect them. For tables with many pages that's not very usable.

enrique-prado commented on src/components/Table/TableComponents/TableHeader.tsx:
```diff
@@ -69,19 +87,71 @@ export const TableHeader = <T extends Record<string, any>>({
+											</>
+										) : (
+											<>
+												{translations.selectAll} ({rows.length})
```
We also need an menu option to clear all selected rows. See my comments about UX in this PR.

joe-s-avaya replied:
![2024-07-11 10 49 32](https://github.com/avaya-dux/neo-react-library/assets/55896546/11a0dc62-cd45-4073-b3c5-042635d40fc2)

enrique-prado commented on src/components/Table/helpers/default-data.ts:
```diff
@@ -20,7 +20,10 @@ export const translations: ITableTranslations = {
 		clearSort: "Clear Sort",
 		filterColumn: "Filter Column",
-		selectAll: "select all",
+		clearSelection: "Clear all selections",
```
I never saw this string in all my testing. Is it missing from your logic? Perhaps a bug?

joe-s-avaya replied:
![2024-07-11 10 49 32](https://github.com/avaya-dux/neo-react-library/assets/55896546/11a0dc62-cd45-4073-b3c5-042635d40fc2)

matthewAvaya commented:
- After selecting **Select all items (10)** from the header dropdown, the banner to **clear all selections** is missing. This light blue banner is missing in some of the other instances. 
<img width="1440" alt="Screen Shot 2024-07-10 at 2 34 03 PM" src="https://github.com/avaya-dux/neo-react-library/assets/90473510/2c16983c-fe53-4028-8abc-25fbc4a3d80f">

enrique-prado reviewed (commented):
There seems to be an issue with the "Clear all selections" option. All my comments are related to that.

joe-s-avaya commented:
@matthewAvaya 

> * After selecting **Select all items (10)** from the header dropdown, the banner to **clear all selections** is missing. This light blue banner is displayed anytime a user selects one of the dropdown items from the header dropdown of the checkbox.

This is not included in this PR. It will be added in the next one. 

> * Change copy **Select all items** to **Select all items in the table**

Updated

> * Change copy **Clear all selections** to **Clear all items**

Updated

yangAtSpoken reviewed (approved):
Looks good to me.

matthewAvaya commented:
> @matthewAvaya
> 
> > * After selecting **Select all items (10)** from the header dropdown, the banner to **clear all selections** is missing. This light blue banner is displayed anytime a user selects one of the dropdown items from the header dropdown of the checkbox.
> 
> This is not included in this PR. It will be added in the next one.
> 
> > * Change copy **Select all items** to **Select all items in the table**
> 
> Updated
> 
> > * Change copy **Clear all selections** to **Clear all items**
> 
> Updated

I like the new copy that you added to the dropdown. 
<img width="309" alt="Screen Shot 2024-07-11 at 10 32 55 AM" src="https://github.com/avaya-dux/neo-react-library/assets/90473510/40fd9504-2b9d-42e2-bc4a-eee2cb563163">

For clearing, let's use a similar copy. 

- Clear all items on page (5)
- Clear all items in the table (10)
